pyclam/manifold.py

```diff
@@ -426,683 +426,380 @@ def from_json(manifold, data):
         if 'str' not in self.cache:
             self.cache['str'] = ', '.join(sorted(list(map(str, self.clusters))))
+        else:
+            pass
         return self.cache['str']
 
     def __repr__(self) -> str:
-        # Cashing value because sort can be expensive on many clusters.
         if 'repr' not in self.cache:
             self.cache['repr'] = '\n'.join(sorted(list(map(repr, self.clusters))))
+        else:
+            pass
         return self.cache['repr']
 
-    def __hash__(self):
+    def __hash__(self) -> int:
         return hash(str(self))
 
-    def __contains__(self, cluster: 'Cluster') -> bool:
+    def __contains__(self, cluster: Cluster) -> bool:
         return cluster in self.clusters
 
+    @property
+    def edges_dict(self) -> Dict[Cluster, Set[Edge]]:
+        if 'edges_dict' not in self.cache:
+            self.cache['edges_dict'] = {
+                cluster: {edge for edge in self.edges if cluster in edge}
+                for cluster in self.clusters
+            }
+        else:
+            pass
+        return self.cache['edges_dict']
+
     @property
     def cardinality(self) -> int:
-        return len(self.edges)
+        return len(self.clusters)
 
     @property
     def population(self) -> int:
         return sum((cluster.cardinality for cluster in self.clusters))
 
     @property
     def manifold(self) -> 'Manifold':
-        return next(iter(self.edges)).manifold
+        return next(iter(self.clusters)).manifold
 
     @property
     def metric(self) -> Metric:
-        return next(iter(self.edges)).metric
+        return next(iter(self.clusters)).metric
 
     @property
     def depth(self) -> int:
         if 'depth' not in self.cache:
             self.cache['depth'] = max((cluster.depth for cluster in self.clusters))
+        else:
+            pass
         return self.cache['depth']
 
     @property
-    def clusters(self) -> Iterable[Cluster]:
-        return self.edges.keys()
+    def min_depth(self) -> int:
+        if 'min_depth' not in self.cache:
+            self.cache['min_depth'] = min((cluster.depth for cluster in self.clusters))
+        else:
+            pass
+        return self.cache['min_depth']
 
     @property
-    def subsumed_clusters(self) -> Set[Cluster]:
-        """ Set of Clusters subsumed by other clusters. """
-        if 'subsumed_clusters' not in self.cache:
-            self.build_edges()
-        return self.cache['subsumed_clusters']
+    def depth_range(self) -> Tuple[int, int]:
+        return self.min_depth, self.depth
 
     @property
-    def walkable_clusters(self) -> Set[Cluster]:
-        """ Set of Clusters not subsumed by other clusters. """
-        if 'walkable_clusters' not in self.cache:
+    def subgraphs(self) -> Set['Graph']:
+        """ Returns the set of all connected component subgraphs in the Graph. """
+        if not self.is_built:
             self.build_edges()
-        return self.cache['walkable_clusters']
```

Comment:
I don't see how this function returns anything

Reply:
Does line 589 in the newest push clear things up? I'm only guessing at what meant here...